src/api/uploads/MultiputUpload.js

```diff
@@ -0,0 +1,367 @@
+import createWorker from '../../util/uploadsSHA1Worker';
+import MultiputPart, { PART_STATE_UPLOADED, PART_STATE_DIGEST_READY } from './MultiputPart';
+import type { StringAnyMap } from '../../flowTypes';
+
+const DEFAULT_MULTIPUT_CONFIG = {
+    console: false,
+    digestReadahead: 5,
+    initialRetryDelayMs: 5000,
+    maxRetryDelayMs: 60000,
+    parallelism: 5,
+    partSize: null,
+    requestTimeout: 120000,
+    retries: 5
+};
+
+class MultiputUpload extends Base {
+    clientId: ?string;
+    commitRetryCount: number;
+    config: Object;
+    createSessionNumRetriesPerformed: number;
+    destinationFile: ?string;
+    destinationFolder: ?string;
+    ended: boolean;
+    file: File;
+    fileSha1: ?string;
+    firstUnuploadedPartIndex: number;
+    initialFileLastModified: ?string;
+    initialFileSize: number;
+    onCompleted: ?Function;
+    onFailure: ?Function;
+    onProgress: ?Function;
+    partSize: number;
+    parts: Array<MultiputPart>;
+    partsDigestComputing: number;
+    partsDigestReady: number;
+    partsNotStarted: number;
+    partsUploaded: number;
+    partsUploading: number;
+    sessionEndpoints: Object;
+    totalUploadedBytes: number;
+    worker: Worker;
+    sessionId: ?string;
+
+    /**
+     * [constructor]
+     * 
+     * @private
+     * @param {object} options
+     * @param {File} file
+     * @param {string} createSessionUrl
+     * @param {string} [destinationFolder] - Untyped folder id (e.g. no "d_" prefix)
+     * @param {string} [destinationFile] - Untyped file id (e.g. no "f_" prefix)
+     * @param {object} config
+     */
+    constructor(
+        options: Object,
+        file: File,
+        createSessionUrl: string,
+        destinationFolder?: string,
+        destinationFile?: string,
+        config: Object
+    ) {
+        super(options);
+
+        if (destinationFolder !== null && destinationFile !== null) {
+            throw new Error('Both destinationFolder and destinationFile set');
+        }
+        this.file = file;
+
+        // These values are used as part of our (best effort) attempt to abort uploads if we detect
+        // a file change during the upload.
+        this.initialFileSize = this.file.size;
+        this.initialFileLastModified = getFileLastModifiedAsISONoMSIfPossible(this.file);
+
+        this.sessionEndpoints = {
+            createSession: createSessionUrl,
+            uploadPart: null,
+            listParts: null,
+            commit: null,
+            abort: null,
+            logEvent: null
+        };
+        // Set a default logEvent endpoint so that we can report failures from createSession.
+        // This value will be overwritten on successful response from createSession.
+        // This assumes that the createSessionUrl already contains the upload session id.
+        this.sessionEndpoints.logEvent = `${createSessionUrl.replace(/files\/\d+/, 'files')}/log`;
```

Comment:
This endpoint is whitelisted to only the API key used by Amsterdam.  I'm fine opening it up to other first-party Box apps, but it seems like this code may perform uploads from 3rd-party apps?

Reply:
Got it. Looks like the session request always return `log_event` url? Removing this default value